### PR TITLE
fix: don't allow the root package dependencies to be overwritten in the map

### LINKED ISSUE #44: Handle rewrite map collisions
When building the rewrite map (`buildRewriteMapping`) new entries for the same `DvcsImport` will overwrite the old ones, so if a `package.json` explicitly declares a dependency A with version 2.0 but another dependency B (processed later) also has a dependency to A but with an older 1.0 version, this older version will get rewritten, the root package will be using the old version even though it explicitly declared the new one.

A first approach would be to prioritize dependencies in the `package.json` (over indirect dependencies), but I also have the doubt of why the map needs to go through the entire dependency tree to rewrite just the root package (I'm thinking of the `gx install` scenario, where the rewrite -`post-install` hook- will be  called explicitly on every package separately).

### PR DESCRIPTION
**Pull request by schomatis**
```
fix: don't allow the root package dependencies to be overwritten in the map

The rewrite map is built recursively scanning the root package dependencies. If
two packages in the tree imported the same dependency with different hash the
last one to be processed would overwrite the first one.

Include a flag in `buildRewriteMapping` to signal when the root package
dependencies are being process and allow them to overwrite the map (adding an
overwrite flag in `addRewriteForDep`) but don't allow the inverse (transitive
dependencies overwriting direct ones).

This change has as a side effect that in a scenario where transitive
dependencies with different versions were overwriting each other that won't be
allowed now, changing in fact the rewrite map, but that would happen only in the
already problematic scenario of dependencies with mismatched versions.
```

-------------------

Fixes #44.

Tested locally, it feels that it should have a test but I don't see any testing framework here (it would be nice to have it to easily create dependencies trees and test different scenarios, I'll open an issue about it.)

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -1005,21 +1005,39 @@ func loadDep(dep *gx.Dependency, pkgdir string) (*Package, error) {
...
+	to := "gx/ipfs/" + dep.Hash + "/" + pkg.Name
+	if undo {
+		from, to = to, from
+	}
+
+	_, entryExists := m[from]
+	if !entryExists || overwrite {
 		m[from] = to
+	} else if entryExists && m[from] != to {
+		VLog("trying to overwrite rewrite map entry %s poiting to %s with %s", from, m[from], to)
```
Comment (schomatis) [inline main.go]: Fix typo: "poiting".

**Comment by Stebalien** [comment]
So, this is *a* fix but, after thinking about this, I think the better fix is to change the traversal order to breadth first (and ban *all* overwriting). This catches the case where the *root* package overrides a dependency version but doesn't apply the same logic recursively.

How about:

1. Adding rewrites for all dependencies of the current package in one loop.
2. Recursing in a second loop.

**Comment by schomatis** [comment]
Do you mind merging this anyway while we discuss this other solution in the original issue? (The other `gx-go link` PR I'm working on already depends on this one, and although not optimal it's still a valid solution that would allow me to concentrate on that before revisiting this one.)

**Comment by Stebalien** [comment]
Fair enough. This fixes the immediate issue.

**Pull request merged**